util/rate_limiter.cc

```diff
@@ -228,6 +240,30 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
   } while (!r.granted);
 }
 
+std::vector<Env::IOPriority>
+GenericRateLimiter::GeneratePriorityIterationOrder() {
```

Comment:
The logic of deciding the order of iterating request queues of different priorities (i.e, the order of satisfying requests of different priorities) is as the following:
(1)There is 1/fairness chance that user_pri requests will be satisfied AFTER BOTH high_pri and low_pri requests. Otherwise, user_pri requests will be satisfied BEFORE BOTH high_pri and low_pri requests.
(2) Regardless of the order of satisfying user_pri requests relative to BOTH high_pri and low_pri, there is 1/fairness chance that high_pri requests will be satisfied AFTER low_pri requests. Otherwise, high_pri requests will be satisfied BEFORE low_pri requests.

Therefore the only 4 possible iteration orders as following:
[user_pri, high_pri, low_pri], [user_pri, low_pri, high_pri], [high_pri, low_pri, user_pri], [low_pri, high_pri, user_pri],

Reply:
[TODO]The current impl verbosely manipulating pri in generating the order for the benefit of readability. Should we have more pri levels, this verbose impl becomes inconvenient to change. It might be worthwhile to think about a less verbose way and impl it now.

Reply:
So once we have user-pri and mid-pri, are these the new iteration orders?

{user_pri, high_pri, mid_pri, low_pri} with probability (1 - 1/fairness) * (1 - 1/fairness)
{user_pri, high_pri, low_pri, mid_pri} with probability (1 - 1/fairness) * 1/fairness
{user_pri, mid_pri, low_pri, high_pri} with probability 1/fairness * (1 - 1/fairness)
{user_pri, low_pri, mid_pri, high_pri} with probability 1/fairness^2

If user_pri and mid_pri are empty we get high_pri above low_pri at (1 - 1/fairness) * (1 - 1/fairness) + (1 - 1/fairness) * 1/fairness = 1 - 1/fairness. And low_pri above high_pri at 1/fairness * (1 - 1/fairness) + 1/fairness^2 = 1/fairness. So it is compatible.

Reply:
>So once we have user-pri and mid-pri, are these the new iteration orders?
Yes - the iteration order and the probability is correct :)

>.. we get high_pri above low_pri at (1 - 1/fairness) * (1 - 1/fairness) + (1 - 1/fairness) * 1/fairness = 1 - 1/fairness. And low_pri above high_pri at 1/fairness * (1 - 1/fairness) + 1/fairness^2 = 1/fairness. So it is compatible.

I agree with the math and the conclusion that it is compatible - just one more note: even when user_pri or mid_pri is not empty, the probability of high_pri above low and low_pri above high stay the same as your calculation. It is because whether user_pri/mid_pri is empty or not is irrelevant to my order generation logic. We generate the order first (as if we are going to have all these pri) and then we iterate the queues according to this order. When a certain pri queue is empty, we just iterate the empty queue as if it is skipped.

(Just in case you wonder, I'd like to think that the reason why we get such a compatible probability result is because we maintain the definition of fairness for a certain priority level P being "one of # times that P will be placed after ALL P's lower pri levels", both in the case where we have only high/low pri or where we have more.)

Reply:
>[TODO]Regardless of which way we are considering the relative order of user_pri to lower pri requests, I need to update the public document and explain how fairness works for user_pri/high_pri/low_pri.

Discussed offline that we are going to go with the current algorithm of deciding order.

>[TODO]The current impl verbosely manipulating pri in generating the order for the benefit of readability. Should we have more pri levels, this verbose impl becomes inconvenient to change. Will consider thinking about a less verbose impl.

Discussed offline that no need for this optimization now since we expect the current amount of priority levels are enough and current verbose impl is good enough for now. If optimization is needed for the future, we can consider rewrite this as recursion.